Event Timing: event counting (https://w3c.github.io/event-timing)

francovs opened the pull request:
#### d494a31e08517348d6cba1659a4613ca1bdd1238
<pre>
Event Timing: event counting (<a href="https://w3c.github.io/event-timing)">https://w3c.github.io/event-timing)</a>
<a href="https://webkit.org/b/296383">https://webkit.org/b/296383</a>
<a href="https://rdar.apple.com/156500312">rdar://156500312</a>

Reviewed by Ryosuke Niwa.

Implements event counting (performance.eventCounts), which is part of
the event timing API.

The way this code implements a maplike interface is problematic: when
the eventCounts object is accessed through javascript, a second
object containing the event counts is created. It is not kept in sync
after that. See `EventCounts::add()` and
`EventCounts::initializeMapLike()`.

This implementation does not count mouse{enter,leave} and
pointer{enter,leave} events if there are no PerformanceObservers
watching them. This happens because an optimization prevents these
events from being dispatched; see
`EventHandler::updateMouseEventTargetNode()`.

* LayoutTests/TestExpectations: skip tests that time out instead of failing early
* LayoutTests/imported/w3c/web-platform-tests/event-timing/event-click-counts-expected.txt: now passes
* LayoutTests/imported/w3c/web-platform-tests/event-timing/event-counts-zero-expected.txt: now passes
* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.window-expected.txt: now passes
* LayoutTests/platform/ios/TestExpectations: skip test that times out instead of failing early
* LayoutTests/platform/mac-wk1/TestExpectations: skip test that times out instead of failing early
* Source/WebCore/platform/Logging.h: add a new logger for the Performance Timeline

Canonical link: <a href="https://commits.webkit.org/298135@main">https://commits.webkit.org/298135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bfb1648fdd1d3187766c6283d7300b3d8b082cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120536 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65091 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86914 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117317 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102713 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67306 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26855 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20839 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64223 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97049 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123745 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41383 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30866 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95741 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41760 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/98910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95525 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24345 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40673 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18510 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37426 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41263 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46771 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40857 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44164 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/EventCounts.cpp:
```diff
@@ -25,14 +25,134 @@
+const AtomString& EventCounts::eventNameFromType(EventType t)
+{
+    static const WTF::NeverDestroyed map(std::map<EventType, const AtomString>({
+        { EventType::auxclick           , eventNames().auxclickEvent           },
```
This and `COUNTED_EVENT_SET` should be auto-generated from EventNames.json (see make-event-names.py)

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2460,48 @@ void LocalDOMWindow::finishedLoading()
+    // FIXME: implement InteractionId logic
+    // event.timeStamp() may not be reliable (wall clock):
+    auto startTime = std::min(event.timeStamp(), MonotonicTime::now());
+    return PerformanceEventTimingCandidate {
```
We shouldn't be creating this object for every event we dispatch.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2460,48 @@ void LocalDOMWindow::finishedLoading()
+    // FIXME: implement InteractionId logic
+    // event.timeStamp() may not be reliable (wall clock):
+    auto startTime = std::min(event.timeStamp(), MonotonicTime::now());
+    return PerformanceEventTimingCandidate {
```
We shouldn't be creating this object for every event we dispatch.

francovs replied:
I believe some of this is unavoidable: the entry's duration is calculated from the event's timestamp until the moment the rendering is updated, but `EventHandler::dispatchEvent()` only knows when the processing of handlers ended (`processingEnd`), which is not enough to be able to discard candidates based on duration.

rniwa replied:
We shouldn't be creating this object upfront. We should create it only when the duration is at least 16ms in finalizeEventTimingEntry. All we need to do in this function should be to compute the start time.

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -176,12 +177,18 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
+    RefPtr window = document->window();
+    std::optional<LocalDOMWindow::PerformanceEventTimingCandidate> pendingEventTiming;
+    if (typeInfo.isInCategory(EventCategory::EventTimingEligible) && window && document->settings().eventTimingEnabled() && event.isTrusted())
+        pendingEventTiming = window->initializeEventTimingEntry(event, typeInfo);
```
This will slow down every event dispatch. That's unacceptable.

francovs replied:
I tried to make this as fast as possible: following @nt1m 's suggestion, the first test (which eliminates most events) now uses `EventCategory`, so it should consist only of applying a bit mask.

Still, I need to check benchmark results to make sure there isn't a significant regression.

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -176,12 +177,18 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
     if (hasNoEventListenerOrDefaultEventHandler && !targetOrRelatedTargetIsInShadowTree) {
         event.resetBeforeDispatch();
         event.setTarget(RefPtr { EventPath::eventTargetRespectingTargetRules(node) });
+        window->finalizeEventTimingEntry(pendingEventTiming, event.target());
```
Instead of manually calling this function whenever we return, use `makeScopeExit` (https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/Scope.h).

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2460,48 @@ void LocalDOMWindow::finishedLoading()
+    LOG_WITH_STREAM(PerformanceTimeline, stream << "Dispatching " << m_performanceEventTimingCandidates.size() << " event timing entries");
+    for (auto &e : m_performanceEventTimingCandidates) {
+        performance().eventCounts()->add(e.typeInfo.type());
+        // FIXME: calculate duration and ignore if < 16ms
```
We should do this check in `LocalDOMWindow::finalizeEventTimingEntry`. We shouldn't be appending a new entry in the first place.

francovs replied:
This wouldn't work with the spec because `duration` is not known at `LocalDOMWindow::finalizeEventTimingEntry`. The spec requires 4 separate timestamps:

1) `startTime`, which is the event's own timestamp
2) `processingStart`, from right before the event handlers are called. This is supposed to be a replacement for JS code adding a `performance.now()` call at the beginning of the first event handler
3) `processingEnd`, taken at the end of event dispatch. This is supposed to be a replacement for JS code calling `performance.now()` at the end of the last event handler. This does not determine the duration, so it cannot be used to discard entries
4) `renderingTimestamp`, taken after rendering is updated. This is exposed through JS as `duration`, which equals `renderingTimestamp - startTime`, with 8ms granularity. Reduced granularity is used to avoid introducing a new fingerprinting channel: there is no way to plug in JS code to obtain `renderingTimestamp` 

According to the spec, this is useful for quantifying separate delays: `processingStart` measures "input delay", `processingEnd` measures "processing delay" and `renderingTimestamp` measures "presentation delay".

rniwa replied:
I see. In that case, we really need to thoroughly test this PR with Speedometer on our perf A/B testing infrastructure.

francovs replied:
Testing shows no significant regression on Speedometer with this PR. However, this PR doesn't exercise the vector because the feature flag is off.

Enabling the flags didn't move the Speedometer needle, but this could be because there is no trusted user input to cause entries to be created.

Surprisingly, enabling the flag slightly progressed page load - @rreno speculates this could be caused by avoiding polyfills.

rniwa commented on Source/WebCore/bindings/js/JSDOMMapLike.h:
```diff
@@ -62,6 +62,7 @@ class DOMMapAdapter {
 private:
     JSC::JSGlobalObject& m_lexicalGlobalObject;
     JSC::JSObject& m_backingMap;
+    friend class EventCounts;
```
Definitely don't do this.

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -176,6 +179,15 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
+    if (typeInfo.isInCategory(EventCategory::EventTimingEligible) && window && document->settings().eventTimingEnabled() && event.isTrusted())
+        pendingEventTiming = window->initializeEventTimingEntry(event, typeInfo);
+
+    ScopeExit<Function<void(void)>> finalizeEntry( [&]() {
```
Do this instead:
```cpp
auto finalizeEntry = makeScopeExit([&] {
    ...
})
```

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -206,6 +218,7 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
     if (hasNoEventListenerOrDefaultEventHandler) {
         if (shouldClearTargetsAfterDispatch)
             resetAfterDispatchInShadowTree(event);
+
```
Revert this.

rniwa commented on Source/WebCore/page/EventCounts.cpp:
```diff
@@ -25,14 +25,41 @@
+        map.set<IDLDOMString, IDLUnsignedLongLong>(allEventNames[typeAsIndex], m_counts[index]);
+    }
+
+    m_maplike = &map.m_backingMap;
```
This is not okay.

rniwa commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -26,21 +26,35 @@
+
+    // FIXME: the maplike IDL interface should read directly
+    // from m_counts instead of doing this
+    JSC::JSGlobalObject* m_maplikeGlobalObject;
```
I don't think this is safe or correct. We can't store a raw pointer to JSGlobalObject/JSObject like this.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2460,48 @@ void LocalDOMWindow::finishedLoading()
+        .typeInfo = typeInfo,
+        .cancelable = event.cancelable(),
+        .startTime =  performance().relativeTimeFromTimeOriginInReducedResolution(startTime),
+        .processingStart = performance().now(),
```
This will invoke `MonotonicTime::now` again inside for no good reason. We should be calling `MonotonicTime::now()` exactly once.

rniwa commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -26,21 +26,35 @@
 
 class DOMMapAdapter;
 
 class EventCounts final: public RefCounted<EventCounts> {
```
Why is this ref-counted at all? There is no shared ownership here.
We should be forwarding `ref`/`deref` to `Performance` instead.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2460,48 @@ void LocalDOMWindow::finishedLoading()
+        return;
+
+    LOG_WITH_STREAM(PerformanceTimeline, stream << "Dispatching " << m_performanceEventTimingCandidates.size() << " event timing entries");
+    for (auto &e : m_performanceEventTimingCandidates) {
```
Please don't use a single letter variable name like "e". Spell out "event".

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -283,6 +285,21 @@ class LocalDOMWindow final
 
+    // EventTiming API
+    struct PerformanceEventTimingCandidate {
+
```
Remove this blank inline.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -283,6 +285,21 @@ class LocalDOMWindow final
+    struct PerformanceEventTimingCandidate {
+
+        EventTypeInfo typeInfo;
+        // Currently unused, but required for full event timing implementation:
```
I don't think this comment is necessary.

rniwa commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -26,21 +26,28 @@
 
 private:
-    EventCounts();
+    Performance* m_container;
```
Please use WeakRef instead.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -283,6 +285,19 @@ class LocalDOMWindow final
+    // EventTiming API
+    struct PerformanceEventTimingCandidate {
+        EventTypeInfo typeInfo;
+        bool cancelable;
```
Initialize it to `false` with `{ false };`

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -283,6 +285,19 @@ class LocalDOMWindow final
+    struct PerformanceEventTimingCandidate {
+        EventTypeInfo typeInfo;
+        bool cancelable;
+        DOMHighResTimeStamp startTime;
```
Initialize these to 0 with `{ 0 };`

rniwa commented on Source/WebCore/page/Performance.h:
```diff
@@ -143,6 +147,7 @@ class Performance final : public RefCounted<Performance>, public ContextDestruct
     void queueEntry(PerformanceEntry&);
     void scheduleTaskIfNeeded();
 
+    mutable std::unique_ptr<EventCounts> m_eventCounts;
```
Use `const std::unique_ptr<EventCounts> m_eventCounts;` and `lazyInitialize`.

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -176,6 +178,15 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
+        pendingEventTiming = window->initializeEventTimingEntry(event, typeInfo);
+    auto finalizeEntry(WTF::makeScopeExit([&] {
+        if (pendingEventTiming)
+            window->finalizeEventTimingEntry(pendingEventTiming, event.target());
```
We need to either capture event using Ref/RefPtr or suppress the lambda capture warning for this lambda.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2459,48 @@ void LocalDOMWindow::finishedLoading()
+    };
+}
+
+void LocalDOMWindow::finalizeEventTimingEntry(const std::optional<PerformanceEventTimingCandidate>& entry, RefPtr<EventTarget> target)
```
This should just take PerformanceEventTimingCandidate& instead.

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -176,6 +178,15 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
+        pendingEventTiming = window->initializeEventTimingEntry(event, typeInfo);
+    auto finalizeEntry(WTF::makeScopeExit([&, protectedTarget=event.protectedTarget()] {
+        if (pendingEventTiming)
+            window->finalizeEventTimingEntry(pendingEventTiming, WTFMove(protectedTarget));
```
We should be passing `*pendingEventTiming` here since we've already checked for `std::nullopt`.

rniwa commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -26,21 +26,29 @@
-    static Ref<EventCounts> create() { return adoptRef(*new EventCounts()); }
+    EventCounts(Performance* container);
+    // Forward refcounting to container:
+    void ref();
```
We can just inline these functions and get rid of the above comment.

rniwa commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -442,6 +457,9 @@ class LocalDOMWindow final
     mutable RefPtr<CloseWatcherManager> m_closeWatcherManager;
 
+    // Equivalent to the list of PerformanceEventTiming objects mentioned in https://www.w3.org/TR/event-timing/#sec-modifications-HTML :
+    Vector<PerformanceEventTimingCandidate> m_performanceEventTimingCandidates;
```
We should consider adding an inline capacity to this `Vector`.

francovs replied:
I used 6, which results in 240 bytes total. Just a wild guess, though.

When testing with logging on (the `PerformanceTimeline` logger), 4 events seem to be, by far, the most common number of entries found when emptying the vector. This is due to the `pointerover`+`mouseover` + `pointerout`+`mouseout` combo, which shows up when moving the pointer across elements.

rniwa commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -260,6 +275,17 @@ void Performance::appendBufferedEntriesByType(const String& entryType, Vector<Re
+    // FIXME: stop lazy-initializing m_eventCounts after event
+    // timing stops being gated by a flag:
+    if (!m_eventCounts)
+        lazyInitialize(m_eventCounts, makeUniqueWithoutRefCountedCheck<EventCounts>(this));
```
Instead of duplicating code like this, we should simply call `eventCounts()`.

rniwa commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2457,6 +2459,46 @@ void LocalDOMWindow::finishedLoading()
+        .startTime =  std::min(processingStart, performance().relativeTimeFromTimeOriginInReducedResolution(event.timeStamp())),
+        .processingStart = processingStart,
+        .processingEnd = 0,
+        .target = nullptr
```
No need to explicitly set processingEnd or target here. They are initialized to 0 & nullptr already.

francovs replied:
Done, but I had to be explicit about initializing `target` to avoid breaking the CMake builds:

```FAILED: Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp.o 
/usr/bin/ccache /usr/bin/c++ -DBUILDING_GTK__=1 -DBUILDING_WEBKIT=1 -DBUILDING_WITH_CMAKE=1 -DBUILDING_WebCore -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DGETTEXT_PACKAGE=\"WebKitGTK-6.0\" -DHAVE_CONFIG_H=1 -DJSC_GLIB_API_ENABLED -DPAS_BMALLOC=1 -DSK_ASSUME_GL=0 -DSK_ASSUME_GL_ES=1 -DSK_ASSUME_WEBGL=0 -DSK_DISABLE_LEGACY_GL_MAKE_NATIVE_INTERFACE -DSK_DISABLE_LEGACY_IMAGE_READBUFFER -DSK_DISABLE_LEGACY_INIT_DECODERS -DSK_DISABLE_LEGACY_PNG_WRITEBUFFER -DSK_DISABLE_TRACING -DSK_DISABLE_WEBGL_INTERFACE -DSK_ENABLE_PRECOMPILE -DSK_R32_SHIFT=16 -DSK_RELEASE -DSTATICALLY_LINKED_WITH_PAL -DTHUNDER_HAS_OCDM_DECRYPT_BUFFER=0 -DUSE_SYSTEM_EGL -D_GLIBCXX_ASSERTIONS=1 -I/app/webkit/WebKitBuild/GTK/Release -I/app/webkit/WebKitBuild/GTK/Release/WebCore/DerivedSources -I/app/webkit/Source/WebCore -I/app/webkit/Source/WebCore/Modules/ShapeDetection -I/app/webkit/Source/WebCore/Modules/ShapeDetection/Interfaces -I/app/webkit/Source/WebCore/Modules/WebGPU -I/app/webkit/Source/WebCore/Modules/WebGPU/InternalAPI -I/app/webkit/Source/WebCore/Modules/WebGPU/Implementation -I/app/webkit/Source/WebCore/Modules/airplay -I/app/webkit/Source/WebCore/Modules/applepay -I/app/webkit/Source/WebCore/Modules/applepay/paymentrequest -I/app/webkit/Source/WebCore/Modules/applicationmanifest -I/app/webkit/Source/WebCore/Modules/async-clipboard -I/app/webkit/Source/WebCore/Modules/audiosession -I/app/webkit/Source/WebCore/Modules/badge -I/app/webkit/Source/WebCore/Modules/beacon -I/app/webkit/Source/WebCore/Modules/cache -I/app/webkit/Source/WebCore/Modules/compression -I/app/webkit/Source/WebCore/Modules/contact-picker -I/app/webkit/Source/WebCore/Modules/cookie-consent -I/app/webkit/Source/WebCore/Modules/cookie-store -I/app/webkit/Source/WebCore/Modules/credentialmanagement -I/app/webkit/Source/WebCore/Modules/encryptedmedia -I/app/webkit/Source/WebCore/Modules/encryptedmedia/legacy -I/app/webkit/Source/WebCore/Modules/entriesapi -I/app/webkit/Source/WebCore/Modules/fetch -I/app/webkit/Source/WebCore/Modules/filesystem -I/app/webkit/Source/WebCore/Modules/geolocation -I/app/webkit/Source/WebCore/Modules/highlight -I/app/webkit/Source/WebCore/Modules/identity -I/app/webkit/Source/WebCore/Modules/identity/dummy -I/app/webkit/Source/WebCore/Modules/identity/protocols -I/app/webkit/Source/WebCore/Modules/identity/protocols/ISO18013 -I/app/webkit/Source/WebCore/Modules/identity/protocols/openid -I/app/webkit/Source/WebCore/Modules/indexeddb -I/app/webkit/Source/WebCore/Modules/indexeddb/client -I/app/webkit/Source/WebCore/Modules/indexeddb/server -I/app/webkit/Source/WebCore/Modules/indexeddb/shared -I/app/webkit/Source/WebCore/Modules/mediacapabilities -I/app/webkit/Source/WebCore/Modules/mediacontrols -I/app/webkit/Source/WebCore/Modules/mediarecorder -I/app/webkit/Source/WebCore/Modules/mediasession -I/app/webkit/Source/WebCore/Modules/mediasource -I/app/webkit/Source/WebCore/Modules/mediastream -I/app/webkit/Source/WebCore/Modules/model-element -I/app/webkit/Source/WebCore/Modules/model-element/dummy -I/app/webkit/Source/WebCore/Modules/navigatorcontentutils -I/app/webkit/Source/WebCore/Modules/notifications -I/app/webkit/Source/WebCore/Modules/paymentrequest -I/app/webkit/Source/WebCore/Modules/permissions -I/app/webkit/Source/WebCore/Modules/pictureinpicture -I/app/webkit/Source/WebCore/Modules/plugins -I/app/webkit/Source/WebCore/Modules/push-api -I/app/webkit/Source/WebCore/Modules/remoteplayback -I/app/webkit/Source/WebCore/Modules/reporting -I/app/webkit/Source/WebCore/Modules/screen-wake-lock -I/app/webkit/Source/WebCore/Modules/speech -I/app/webkit/Source/WebCore/Modules/storage -I/app/webkit/Source/WebCore/Modules/streams -I/app/webkit/Source/WebCore/Modules/url-pattern -I/app/webkit/Source/WebCore/Modules/web-locks -I/app/webkit/Source/WebCore/Modules/webaudio -I/app/webkit/Source/WebCore/Modules/webauthn -I/app/webkit/Source/WebCore/Modules/webauthn/cbor -I/app/webkit/Source/WebCore/Modules/webauthn/fido -I/app/webkit/Source/WebCore/Modules/webcodecs -I/app/webkit/Source/WebCore/Modules/webdatabase -I/app/webkit/Source/WebCore/Modules/webdriver -I/app/webkit/Source/WebCore/Modules/websockets -I/app/webkit/Source/WebCore/Modules/webtransport -I/app/webkit/Source/WebCore/Modules/webxr -I/app/webkit/Source/WebCore/accessibility -I/app/webkit/Source/WebCore/accessibility/isolatedtree -I/app/webkit/Source/WebCore/animation -I/app/webkit/Source/WebCore/automation -I/app/webkit/Source/WebCore/bindings -I/app/webkit/Source/WebCore/bindings/js -I/app/webkit/Source/WebCore/bridge -I/app/webkit/Source/WebCore/bridge/c -I/app/webkit/Source/WebCore/bridge/jsc -I/app/webkit/Source/WebCore/contentextensions -I/app/webkit/Source/WebCore/crypto -I/app/webkit/Source/WebCore/crypto/algorithms -I/app/webkit/Source/WebCore/crypto/keys -I/app/webkit/Source/WebCore/crypto/parameters -I/app/webkit/Source/WebCore/css -I/app/webkit/Source/WebCore/css/calc -I/app/webkit/Source/WebCore/css/parser -I/app/webkit/Source/WebCore/css/query -I/app/webkit/Source/WebCore/css/typedom -I/app/webkit/Source/WebCore/css/typedom/color -I/app/webkit/Source/WebCore/css/typedom/numeric -I/app/webkit/Source/WebCore/css/typedom/transform -I/app/webkit/Source/WebCore/css/values -I/app/webkit/Source/WebCore/css/values/borders -I/app/webkit/Source/WebCore/css/values/color -I/app/webkit/Source/WebCore/css/values/color-adjust -I/app/webkit/Source/WebCore/css/values/easing -I/app/webkit/Source/WebCore/css/values/filter-effects -I/app/webkit/Source/WebCore/css/values/grid -I/app/webkit/Source/WebCore/css/values/images -I/app/webkit/Source/WebCore/css/values/motion -I/app/webkit/Source/WebCore/css/values/primitives -I/app/webkit/Source/WebCore/css/values/shapes -I/app/webkit/Source/WebCore/css/values/text-decoration -I/app/webkit/Source/WebCore/cssjit -I/app/webkit/Source/WebCore/dom -I/app/webkit/Source/WebCore/dom/messageports -I/app/webkit/Source/WebCore/domjit -I/app/webkit/Source/WebCore/editing -I/app/webkit/Source/WebCore/fileapi -I/app/webkit/Source/WebCore/history -I/app/webkit/Source/WebCore/html -I/app/webkit/Source/WebCore/html/canvas -I/app/webkit/Source/WebCore/html/closewatcher -I/app/webkit/Source/WebCore/html/forms -I/app/webkit/Source/WebCore/html/parser -I/app/webkit/Source/WebCore/html/shadow -I/app/webkit/Source/WebCore/html/track -I/app/webkit/Source/WebCore/inspector -I/app/webkit/Source/WebCore/inspector/agents -I/app/webkit/Source/WebCore/inspector/agents/page -I/app/webkit/Source/WebCore/inspector/agents/worker -I/app/webkit/Source/WebCore/layout -I/app/webkit/Source/WebCore/layout/formattingContexts -I/app/webkit/Source/WebCore/layout/formattingContexts/block -I/app/webkit/Source/WebCore/layout/formattingContexts/block/tablewrapper -I/app/webkit/Source/WebCore/layout/formattingContexts/flex -I/app/webkit/Source/WebCore/layout/floats -I/app/webkit/Source/WebCore/layout/formattingContexts/inline -I/app/webkit/Source/WebCore/layout/formattingContexts/inline/display -I/app/webkit/Source/WebCore/layout/formattingContexts/inline/invalidation -I/app/webkit/Source/WebCore/layout/formattingContexts/inline/ruby -I/app/webkit/Source/WebCore/layout/formattingContexts/inline/text -I/app/webkit/Source/WebCore/layout/integration -I/app/webkit/Source/WebCore/layout/integration/flex -I/app/webkit/Source/WebCore/layout/integration/inline -I/app/webkit/Source/WebCore/layout/layouttree -I/app/webkit/Source/WebCore/layout/formattingContexts/table -I/app/webkit/Source/WebCore/loader -I/app/webkit/Source/WebCore/loader/appcache -I/app/webkit/Source/WebCore/loader/archive -I/app/webkit/Source/WebCore/loader/archive/mhtml -I/app/webkit/Source/WebCore/loader/cache -I/app/webkit/Source/WebCore/loader/icon -I/app/webkit/Source/WebCore/mathml -I/app/webkit/Source/WebCore/page -I/app/webkit/Source/WebCore/page/csp -I/app/webkit/Source/WebCore/page/scrolling -I/app/webkit/Source/WebCore/page/text-extraction -I/app/webkit/Source/WebCore/platform -I/app/webkit/Source/WebCore/platform/animation -I/app/webkit/Source/WebCore/platform/audio -I/app/webkit/Source/WebCore/platform/calc -I/app/webkit/Source/WebCore/platform/encryptedmedia -I/app/webkit/Source/WebCore/platform/gamepad -I/app/webkit/Source/WebCore/platform/graphics -I/app/webkit/Source/WebCore/platform/graphics/cpu/arm -I/app/webkit/Source/WebCore/platform/graphics/cpu/arm/filters -I/app/webkit/Source/WebCore/platform/graphics/displaylists -I/app/webkit/Source/WebCore/platform/graphics/filters -I/app/webkit/Source/WebCore/platform/graphics/filters/software -I/app/webkit/Source/WebCore/platform/graphics/controls -I/app/webkit/Source/WebCore/platform/graphics/iso -I/app/webkit/Source/WebCore/platform/graphics/opentype -I/app/webkit/Source/WebCore/platform/graphics/transforms -I/app/webkit/Source/WebCore/platform/mediacapabilities -I/app/webkit/Source/WebCore/platform/mediarecorder -I/app/webkit/Source/WebCore/platform/mediastream -I/app/webkit/Source/WebCore/platform/mock -I/app/webkit/Source/WebCore/platform/mock/mediasource -I/app/webkit/Source/WebCore/platform/network -I/app/webkit/Source/WebCore/platform/sql -I/app/webkit/Source/WebCore/platform/text -I/app/webkit/Source/WebCore/platform/xr -I/app/webkit/Source/WebCore/plugins -I/app/webkit/Source/WebCore/rendering -I/app/webkit/Source/WebCore/rendering/line -I/app/webkit/Source/WebCore/rendering/mathml -I/app/webkit/Source/WebCore/rendering/shapes -I/app/webkit/Source/WebCore/rendering/style -I/app/webkit/Source/WebCore/rendering/svg -I/app/webkit/Source/WebCore/rendering/svg/legacy -I/app/webkit/Source/WebCore/rendering/updating -I/app/webkit/Source/WebCore/replay -I/app/webkit/Source/WebCore/storage -I/app/webkit/Source/WebCore/style -I/app/webkit/Source/WebCore/style/values -I/app/webkit/Source/WebCore/style/values/align -I/app/webkit/Source/WebCore/style/values/anchor-position -I/app/webkit/Source/WebCore/style/values/backgrounds -I/app/webkit/Source/WebCore/style/values/borders -I/app/webkit/Source/WebCore/style/values/box -I/app/webkit/Source/WebCore/style/values/color-adjust -I/app/webkit/Source/WebCore/style/values/color -I/app/webkit/Source/WebCore/style/values/contain -I/app/webkit/Source/WebCore/style/values/content -I/app/webkit/Source/WebCore/style/values/counter-styles -I/app/webkit/Source/WebCore/style/values/easing -I/app/webkit/Source/WebCore/style/values/fill-stroke -I/app/webkit/Source/WebCore/style/values/filter-effects -I/app/webkit/Source/WebCore/style/values/flexbox -I/app/webkit/Source/WebCore/style/values/grid -I/app/webkit/Source/WebCore/style/values/images -I/app/webkit/Source/WebCore/style/values/inline -I/app/webkit/Source/WebCore/style/values/line-grid -I/app/webkit/Source/WebCore/style/values/lists -I/app/webkit/Source/WebCore/style/values/masking -I/app/webkit/Source/WebCore/style/values/motion -I/app/webkit/Source/WebCore/style/values/non-standard -I/app/webkit/Source/WebCore/style/values/overflow -I/app/webkit/Source/WebCore/style/values/position -I/app/webkit/Source/WebCore/style/values/primitives -I/app/webkit/Source/WebCore/style/values/rhythm -I/app/webkit/Source/WebCore/style/values/scroll-animations -I/app/webkit/Source/WebCore/style/values/scroll-snap -I/app/webkit/Source/WebCore/style/values/scrollbars -I/app/webkit/Source/WebCore/style/values/shapes -I/app/webkit/Source/WebCore/style/values/size-adjust -I/app/webkit/Source/WebCore/style/values/sizing -I/app/webkit/Source/WebCore/style/values/svg -I/app/webkit/Source/WebCore/style/values/text -I/app/webkit/Source/WebCore/style/values/text-decoration -I/app/webkit/Source/WebCore/style/values/transforms -I/app/webkit/Source/WebCore/style/values/ui -I/app/webkit/Source/WebCore/style/values/view-transitions -I/app/webkit/Source/WebCore/svg -I/app/webkit/Source/WebCore/svg/animation -I/app/webkit/Source/WebCore/svg/graphics -I/app/webkit/Source/WebCore/svg/graphics/filters -I/app/webkit/Source/WebCore/svg/properties -I/app/webkit/Source/WebCore/websockets -I/app/webkit/Source/WebCore/workers -I/app/webkit/Source/WebCore/workers/service -I/app/webkit/Source/WebCore/workers/service/background-fetch -I/app/webkit/Source/WebCore/workers/service/context -I/app/webkit/Source/WebCore/workers/service/server -I/app/webkit/Source/WebCore/workers/shared -I/app/webkit/Source/WebCore/workers/shared/context -I/app/webkit/Source/WebCore/worklets -I/app/webkit/Source/WebCore/xml -I/app/webkit/Source/WebCore/xml/parser -I/app/webkit/Source/WebCore/Modules/gamepad -I/app/webkit/Source/WebCore/platform/encryptedmedia/clearkey -I/app/webkit/Source/WebCore/platform/graphics/angle -I/app/webkit/Source/WebCore/platform/graphics/lcms -I/app/webkit/Source/ThirdParty/xdgmime/src -I/app/webkit/Source/WebCore/platform/adwaita -I/app/webkit/Source/WebCore/platform/graphics/adwaita -I/app/webkit/Source/WebCore/Modules/mediastream/gstreamer -I/app/webkit/Source/WebCore/platform/graphics/gstreamer -I/app/webkit/Source/WebCore/platform/graphics/gstreamer/mse -I/app/webkit/Source/WebCore/platform/graphics/gstreamer/eme -I/app/webkit/Source/WebCore/platform/graphics/gstreamer/telemetry -I/app/webkit/Source/WebCore/platform/gstreamer -I/app/webkit/Source/WebCore/platform/mediarecorder/gstreamer -I/app/webkit/Source/WebCore/platform/audio/gstreamer -I/app/webkit/Source/WebCore/platform/image-decoders -I/app/webkit/Source/WebCore/platform/image-decoders/avif -I/app/webkit/Source/WebCore/platform/image-decoders/bmp -I/app/webkit/Source/WebCore/platform/image-decoders/gif -I/app/webkit/Source/WebCore/platform/image-decoders/ico -I/app/webkit/Source/WebCore/platform/image-decoders/jpeg -I/app/webkit/Source/WebCore/platform/image-decoders/jpegxl -I/app/webkit/Source/WebCore/platform/image-decoders/png -I/app/webkit/Source/WebCore/platform/image-decoders/webp -I/app/webkit/Source/WebCore/platform/network/soup -I/app/webkit/Source/WebCore/platform/graphics/texmap -I/app/webkit/Source/WebCore/page/scrolling/coordinated -I/app/webkit/Source/WebCore/platform/graphics/texmap/coordinated -I/app/webkit/Source/WebCore/platform/graphics/harfbuzz -I/app/webkit/Source/WebCore/platform/graphics/filters/skia -I/app/webkit/Source/WebCore/platform/graphics/skia -I/app/webkit/Source/WebCore/accessibility/atspi -I/app/webkit/Source/WebCore/crypto/openssl -I/app/webkit/Source/WebCore/page/gtk -I/app/webkit/Source/WebCore/platform/audio/glib -I/app/webkit/Source/WebCore/platform/generic -I/app/webkit/Source/WebCore/platform/glib -I/app/webkit/Source/WebCore/platform/gtk -I/app/webkit/Source/WebCore/platform/graphics/egl -I/app/webkit/Source/WebCore/platform/graphics/epoxy -I/app/webkit/Source/WebCore/platform/graphics/gbm -I/app/webkit/Source/WebCore/platform/graphics/gtk -I/app/webkit/Source/WebCore/platform/graphics/opengl -I/app/webkit/Source/WebCore/platform/graphics/wayland -I/app/webkit/Source/WebCore/platform/graphics/x11 -I/app/webkit/Source/WebCore/platform/mediastream/gtk -I/app/webkit/Source/WebCore/platform/mediastream/gstreamer -I/app/webkit/Source/WebCore/platform/network/glib -I/app/webkit/Source/WebCore/platform/video-codecs -I/app/webkit/Source/WebCore/platform/text/gtk -I/app/webkit/WebKitBuild/GTK/Release/ANGLE/Headers -I/app/webkit/Source/ThirdParty/skia -I/app/webkit/WebKitBuild/GTK/Release/Skia/Headers/top -I/app/webkit/WebKitBuild/GTK/Release/Skia/Headers/mod -I/app/webkit/WebKitBuild/GTK/Release/JavaScriptCore/Headers -I/app/webkit/WebKitBuild/GTK/Release/JavaScriptCore/PrivateHeaders -I/app/webkit/WebKitBuild/GTK/Release/JavaScriptCoreGLib/Headers -I/app/webkit/WebKitBuild/GTK/Release/JavaScriptCoreGLib/DerivedSources -I/app/webkit/WebKitBuild/GTK/Release/WTF/Headers -I/app/webkit/WebKitBuild/GTK/Release/bmalloc/Headers -I/app/webkit/WebKitBuild/GTK/Release/PAL/Headers -isystem /usr/include/gstreamer-1.0 -isystem /usr/include/glib-2.0 -isystem /usr/lib/x86_64-linux-gnu/glib-2.0/include -isystem /usr/include/libdrm -isystem /usr/include/orc-0.4 -isystem /usr/lib/x86_64-linux-gnu/gstreamer-1.0/include -isystem /usr/include/blkid -isystem /usr/include/libmount -isystem /usr/include/WPEFramework/ocdm -isystem /usr/include/flite -isystem /usr/include/libsoup-3.0 -isystem /usr/include/enchant-2 -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/libsecret-1 -isystem /usr/include/libxml2 -isystem /usr/include/avif -isystem /usr/include/harfbuzz -isystem /usr/include/gtk-4.0 -isystem /usr/include/pango-1.0 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/cairo -isystem /usr/include/graphene-1.0 -isystem /usr/lib/x86_64-linux-gnu/graphene-1.0/include -isystem /usr/include/freetype2 -isystem /usr/include/webp -isystem /usr/include/pixman-1 -isystem /usr/include/libpng16 -isystem /usr/include/fribidi -isystem /usr/include/libmanette -isystem /usr/include/sysprof-6 -DASSERT_ENABLED=1 -fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-odr -Wno-stringop-overread -Wno-stringop-overflow -Wno-nonnull -Wno-array-bounds -Wno-expansion-to-defined -Wno-noexcept-type -Wno-subobject-linkage -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fdebug-types-section -fno-omit-frame-pointer -fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++23 -fPIC -fvisibility=hidden -Werror -DAVIF_DLL -mfpmath=sse -msse -msse2 -pthread -MD -MT Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp.o -MF Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp.o.d -o Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp.o -c /app/webkit/WebKitBuild/GTK/Release/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp
In file included from /app/webkit/WebKitBuild/GTK/Release/WebCore/DerivedSources/unified-sources/UnifiedSource-767013ce-5.cpp:7:
/app/webkit/Source/WebCore/page/LocalDOMWindow.cpp: In member function ‘WebCore::LocalDOMWindow::PerformanceEventTimingCandidate WebCore::LocalDOMWindow::initializeEventTimingEntry(const WebCore::Event&, WebCore::EventTypeInfo)’:
/app/webkit/Source/WebCore/page/LocalDOMWindow.cpp:2473:5: error: missing initializer for member ‘WebCore::LocalDOMWindow::PerformanceEventTimingCandidate::target’ [-Werror=missing-field-initializers]
 2473 |     };
      |     ^
cc1plus: all warnings being treated as errors```

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2333,6 +2333,10 @@ void Page::doAfterUpdateRendering()
         document.enqueuePaintTimingEntryIfNeeded();
     });
 
+    forEachDocument([] (Document& document) {
```
This should really be using `forEachRenderableDocument`. The spec refers to the `docs` variable which is filtered before hand in the HTML standard.

nt1m commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -40,6 +40,7 @@
 #include "PointerCaptureController.h"
 #include "UserGestureIndicator.h"
 #include "VoidCallback.h"
+#include <JavaScriptCore/ConsoleTypes.h>
```
This change is redundant

nt1m commented on Source/WebCore/page/Performance.h:
```diff
@@ -34,11 +34,13 @@
 
 #include "ContextDestructionObserver.h"
 #include "DOMHighResTimeStamp.h"
+#include "EventNames.h"
```
I think you should be able to forward declare EventType instead (maybe I'm wrong though) and only import in the CPP

francovs replied:
Done! Forward declaring `EventType` requires specifying the underlying type, which felt bad at first (ODR violation = UB), but on further reflection the underlying type should be quite stable. Also, our build system detects this ODR violation instead of silently accepting it.

rniwa commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -26,21 +26,28 @@
 
 private:
-    EventCounts();
+    WeakRef<Performance, Performance::WeakPtrImplType> m_container;
```
I'd call this variable `m_performance` instead.

rniwa commented on Source/WebCore/page/EventCounts.cpp:
```diff
@@ -26,13 +26,34 @@
+    auto allEventNames = eventNames().allEventNames();
+    for (size_t index = 0; index < EventNames::timedEvents.size(); ++index) {
+        // Subtract 1 to account for EventType::custom
+        size_t typeAsIndex = static_cast<size_t>(EventNames::timedEvents[index]) - 1;
```
Can we add a helper function like `eventNameFromEventType` in `EventNames.h` instead of hard-coding the conversion formula here?

francovs committed:
Event Timing: event counting (https://w3c.github.io/event-timing)
https://webkit.org/b/296383
rdar://156500312

Reviewed by Ryosuke Niwa.

Implements event counting (performance.eventCounts), which is part of
the event timing API.

The way this code implements a maplike interface is problematic: when
the eventCounts object is accessed through javascript, a second
object containing the event counts is created. It is not kept in sync
after that. See `EventCounts::add()` and
`EventCounts::initializeMapLike()`.

This implementation does not count mouse{enter,leave} and
pointer{enter,leave} events if there are no PerformanceObservers
watching them. This happens because an optimization prevents these
events from being dispatched; see
`EventHandler::updateMouseEventTargetNode()`.

* LayoutTests/TestExpectations: skip tests that time out instead of failing early
* LayoutTests/imported/w3c/web-platform-tests/event-timing/event-click-counts-expected.txt: now passes
* LayoutTests/imported/w3c/web-platform-tests/event-timing/event-counts-zero-expected.txt: now passes
* LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.window-expected.txt: now passes
* LayoutTests/platform/ios/TestExpectations: skip test that times out instead of failing early
* LayoutTests/platform/mac-wk1/TestExpectations: skip test that times out instead of failing early
* Source/WebCore/platform/Logging.h: add a new logger for the Performance Timeline

Canonical link: https://commits.webkit.org/298135@main
